core/state_processor.go

```diff
@@ -190,33 +222,43 @@ func (p *StateProcessor) Process(block *types.Block, statedb *state.StateDB, cfg
 			if count < maxMovedCount {
 				count++ // count increase for the account itself
 
-				mkv.addAccount(addr, acc)
-				vkt.ClearStrorageRootConversion(addr)
+				mkv.addAccount(migrdb.GetCurrentAccountAddress().Bytes(), acc)
+				vkt.ClearStrorageRootConversion(migrdb.GetCurrentAccountAddress().Bytes())
 
 				// Store the account code if present
 				if !bytes.Equal(acc.CodeHash, emptyCodeHash[:]) {
 					code := rawdb.ReadCode(statedb.Database().DiskDB(), common.BytesToHash(acc.CodeHash))
 					chunks := trie.ChunkifyCode(code)
 
-					mkv.addAccountCode(addr, uint64(len(code)), chunks)
+					mkv.addAccountCode(migrdb.GetCurrentAccountAddress().Bytes(), uint64(len(code)), chunks)
 				}
 
 				// reset storage iterator marker for next account
-				statedb.Database().SetStorageProcessed(false)
-				statedb.Database().SetCurrentSlotHash(common.Hash{})
+				migrdb.SetStorageProcessed(false)
+				migrdb.SetCurrentSlotHash(common.Hash{})
 
 				// Move to the next account, if available - or end
 				// the transition otherwise.
 				if accIt.Next() {
-					statedb.Database().SetCurrentAccountHash(accIt.Hash())
+					var addr common.Address
+					if _, err := io.ReadFull(fpreimages, addr[:]); err != nil {
+						return nil, nil, 0, fmt.Errorf("reading preimage file: %s", err)
+					}
+					// fmt.Printf("account switch: %s != %s\n", crypto.Keccak256Hash(addr[:]), accIt.Hash())
+					if crypto.Keccak256Hash(addr[:]) != accIt.Hash() {
+						return nil, nil, 0, fmt.Errorf("preimage file does not match account hash: %s != %s", crypto.Keccak256Hash(addr[:]), accIt.Hash())
+					}
+					preimageSeek += int64(len(addr))
+					migrdb.SetCurrentAccountAddress(addr)
 				} else {
 					// case when the account iterator has
 					// reached the end but count < maxCount
-					statedb.Database().EndVerkleTransition()
+					migrdb.EndVerkleTransition()
 					break
 				}
 			}
 		}
+		migrdb.SetCurrentPreimageOffset(preimageSeek)
```

Comment:
When we finish with this block, register the new preimage offset file with what we have been tracking in `preimageSeek`.

This mean that the next block, will open the file at this new offset with, again, means "the next preimage to be resolved".